cpp/src/parquet/column_writer.cc

```diff
@@ -1301,6 +1301,10 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
     bool single_nullable_element =
         (level_info_.def_level == level_info_.repeated_ancestor_def_level + 1) &&
         leaf_field_nullable;
+    if (!leaf_field_nullable && leaf_array.null_count() != 0) {
```

Comment:
Should we take special care when it is `kUnknownNullCount`? It is possible for an end user to produce an array without setting a precise null_count value.

Reply:
I just found that `Array::Validate()` and `Array::ValidateFull()` are checking `DataType` not `Field`. If it knows the latter, we can also validate null_count with nullable from that Field.

Reply:
> Should we take special care when it is kUnknownNullCount? It is possible for an end user to produce an array without setting a precise null_count value.

`kUnknownNullCount` would be counted when it's that value

Reply:
Yes in most cases. I remember that when I did the trick to make mutable (and reusable) arrays. An empty ArrayData is first created by calling `std::shared_ptr<ArrayData> ArrayData::Make(std::shared_ptr<DataType> type, int64_t length, int64_t null_count, int64_t offset)` with default values and then set it to array via `void Array::SetData(const std::shared_ptr<ArrayData>& data)`. The null_count is -1 if not manually maintained.

Reply:
```c++
int64_t ArrayData::GetNullCount() const {
  int64_t precomputed = this->null_count.load();
  if (ARROW_PREDICT_FALSE(precomputed == kUnknownNullCount)) {
    if (this->buffers[0]) {
      precomputed = this->length -
                    CountSetBits(this->buffers[0]->data(), this->offset, this->length);
    } else {
      precomputed = 0;
    }
    this->null_count.store(precomputed);
  }
  return precomputed;
}
```

`null_count()` call will generate null_count if not exists.

Reply:
> I just found that `Array::Validate()` and `Array::ValidateFull()` are checking `DataType` not `Field`. If it knows the latter, we can also validate null_count with nullable from that Field.

See https://github.com/apache/arrow/issues/31387